### ISSUE TITLE
Surprising flow-analysis of `switch` statements

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#4048]
The rule about switch statements is [here](https://github.com/dart-lang/language/blob/master/resources/type-system/flow-analysis.md#statements). Consider the following example:

```dart
enum Enum { one, two }

void main() {
  Enum e = Enum.one;
  num n;
  switch (e) {
    case Enum.one: n = 1; break;
    case Enum.two: n = 2; break;
  }
  print(n.isEven);
}
```

The switch statement rule might need a few adjustments (saying something about `default`, and handling the situation where there is no `default` and the last case has no `break;`), but I would in any case expect the rule to promote `n` to `int` after the switch statement. The actual flow analysis does not reach that conclusion, so the analyzer reports an error on `n.isEven`, showing that `n` has type `num`.

@stereotype441, what do you think?

**Comment by srawlins** [comment]
Agreed. Redirecting to the language repo. Maybe can be found as a duplicate of an existing issue.

**Comment by lrhn** [comment]
Assignment promotion happens for "types of interest", which is the declared type, any type that the variable has been tested against with an `is` or `as` (and probably `case`?), and the nullable/non-nullable alternate of any type of interest.

In this case `int` is not of interest, so assigning an `int` to `n` does not promote.
